internal/ioextensions/reader_with_retrys.go

```diff
@@ -0,0 +1,76 @@
+package ioextensions
+
+import (
+	"io"
+
+	"github.com/wal-g/tracelog"
+)
+
+type ReaderWithRetry struct {
+	reader        io.ReadCloser
+	getReader     func() (io.ReadCloser, error)
+	retryAttempts int
+	alreadyRead   int
+}
+
+func NewReaderWithRetry(getReader func() (io.ReadCloser, error), retryAttempts int) io.ReadCloser {
+	if retryAttempts <= 0 {
+		retryAttempts = 1
+	}
+	return &ReaderWithRetry{
+		reader:        nil,
+		getReader:     getReader,
+		retryAttempts: retryAttempts,
+		alreadyRead:   0,
+	}
+}
+
+func (r *ReaderWithRetry) setupNewReader() error {
+	reader, err := r.getReader()
+	if err != nil {
+		return err
+	}
+	_, err = io.CopyN(io.Discard, reader, int64(r.alreadyRead))
+	r.reader = reader
+	return err
+}
+
+func (r *ReaderWithRetry) Read(p []byte) (int, error) {
+	n := 0
+	for attempt := 0; attempt < r.retryAttempts; attempt++ {
```

Comment:
`Read` method will be called many times with small buffers.
We should make N attempts to download the whole file, not to perform single `Read` call.
It would be better to store `attempt` counter as  `ReaderWithRetry` field.